src/coreclr/vm/array.cpp

```diff
@@ -1210,29 +1210,14 @@ MethodDesc* GetActualImplementationForArrayGenericIListOrIReadOnlyListMethod(Met
     }
     CONTRACTL_END
 
-    int slot = pItfcMeth->GetSlot();
-
-    // We need to pick the right starting method depending on the depth of the inheritance chain
-    static const BinderMethodID startingMethod[] = {
-        METHOD__SZARRAYHELPER__GETENUMERATOR,   // First method of IEnumerable`1
-        METHOD__SZARRAYHELPER__GET_COUNT,       // First method of ICollection`1/IReadOnlyCollection`1
-        METHOD__SZARRAYHELPER__GET_ITEM         // First method of IList`1/IReadOnlyList`1
-    };
-
     // Subtract one for the non-generic IEnumerable that the generic enumerable inherits from
     unsigned int inheritanceDepth = pItfcMeth->GetMethodTable()->GetNumInterfaces() - 1;
-    PREFIX_ASSUME(0 <= inheritanceDepth && inheritanceDepth < ARRAY_SIZE(startingMethod));
-
-    MethodDesc *pGenericImplementor = CoreLibBinder::GetMethod((BinderMethodID)(startingMethod[inheritanceDepth] + slot));
 
-    // The most common reason for this assert is that the order of the SZArrayHelper methods in
-    // corelib.h does not match the order they are implemented on the generic interfaces.
-    _ASSERTE(pGenericImplementor == MemberLoader::FindMethodByName(g_pSZArrayHelperClass, pItfcMeth->GetName()));
+    MethodDesc *pGenericImplementor = MemberLoader::FindMethodByName(g_pSZArrayHelperClass, pItfcMeth->GetName());
 
     // OPTIMIZATION: For any method other than GetEnumerator(), we can safely substitute
     // "Object" for reference-type theT's. This causes fewer methods to be instantiated.
-    if (startingMethod[inheritanceDepth] != METHOD__SZARRAYHELPER__GETENUMERATOR &&
-        !theT.IsValueType())
+    if (inheritanceDepth != 1 && !theT.IsValueType())
```

Comment:
Shouldn't this be `inheritanceDepth != 0`?

My understanding is that previously we had:
* `IList<T> -> ICollection<T> -> IEnumerable<T> -> IEnumerable`
* `IReadOnlyList<T> -> IReadOnlyCollection<T> -> IEnumerable<T> -> IEnumerable`

So in both cases we have a maximum inheritance depth of `4`. We then subtract one so that we're effectively ignoring the existence of `IEnumerable`, since this is only being used for the synthesized generic interfaces.

Since we're still subtracting 1 from the depth above, we want to check against `0` to preserve the optimization.

------------------------

In the new world, we have:
```
         |-> ICollection<T>   -\
IList<T> |                      -> IReadOnlyCollection<T> -> IEnumerable<T> -> IEnumerable
         |-> IReadOnlyList<T> -/
```

So `IEnumerable`, `IEnumerable<T>`, `IReadOnlyCollection<T>`, and `IReadOnlyList<T>` should remain at `-1`, `0`, `1`, and `2` respectively (after the adjustment we're doing on L1214).

While `ICollection<T>` has had its adjusted depth bumped by 1 from `2` to `3`, sharing the slot with `IReadOnlyList<T>`. Similarly, `IList<T>` has been bumped from `3` to `4` remaining a unique slot.

Thus, the lookup optimization can be preserved for any slot other than `3` where `ICollection<T>` and `IReadOnlyList<T>` need something additional to disambiguate them.

Reply:
I pushed up the one line fix here. Tests looked to generally be passing locally with it resolved.

I think we could add back in the lookup optimization as well, but left that out.

Reply:
I did get 1 local failure:
```
Failed test: GC_API::_NoGCRegion_Callback_Svr_Callback_Svr_cmd (GC\\API\\NoGCRegion\\Callback_Svr\\Callback_Svr.cmd) (GC.API.XUnitWrapper.dll)

Assert failure(PID 35420 [0x00008a5c], Thread: 30424 [0x76d8]): promoted_bytes (heap_number) == promoted

CORECLR! SVR::gc_heap::mark_phase + 0x167C (0x00007ff8`5149f56c)
CORECLR! SVR::gc_heap::gc1 + 0x240 (0x00007ff8`514912e0)
CORECLR! SVR::gc_heap::garbage_collect + 0x99B (0x00007ff8`5149104b)
CORECLR! SVR::gc_heap::gc_thread_function + 0x504 (0x00007ff8`514929f4)
CORECLR! SVR::gc_heap::gc_thread_stub + 0x7E (0x00007ff8`51492e9e)
CORECLR! <lambda_0495d23a19b8760473be36722da1494c>::<lambda_invoker_cdecl> + 0x74 (0x00007ff8`5115ca14)
KERNEL32! BaseThreadInitThunk + 0x1D (0x00007ff8`cdc3257d)
NTDLL! RtlUserThreadStart + 0x28 (0x00007ff8`cea8aa58)
    File: D:\Users\tagoo\source\repos\runtime\src\coreclr\gc\gc.cpp:24248
    Image: D:\Users\tagoo\source\repos\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\corerun.exe


Return code:      1
Raw output file:      D:\Users\tagoo\source\repos\runtime\artifacts\tests\coreclr\windows.x64.Checked\Reports\GC.API\NoGCRegion\Callback_Svr\Callback_Svr.output.txt
Raw output:
BEGIN EXECUTION

D:\Users\tagoo\source\repos\runtime\artifacts\tests\coreclr\windows.x64.Checked\GC\API\NoGCRegion\Callback_Svr>set DOTNETCLI=D:\Users\tagoo\source\repos\runtime\dotnet.cmd

D:\Users\tagoo\source\repos\runtime\artifacts\tests\coreclr\windows.x64.Checked\GC\API\NoGCRegion\Callback_Svr>set __DOTNET_ADD_64BIT=1

D:\Users\tagoo\source\repos\runtime\artifacts\tests\coreclr\windows.x64.Checked\GC\API\NoGCRegion\Callback_Svr>set __DOTNET_PREFERRED_BITNESS=64
 "D:\Users\tagoo\source\repos\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\corerun.exe" -p "System.Reflection.Metadata.MetadataUpdater.IsSupported=false" -p "System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization=true"  Callback_Svr.dll
Expected: 100
Actual: -1073740286
```